src/compiler.rs

```diff
@@ -328,9 +328,39 @@ impl<CM: CellManager, SSB: StepSelectorBuilder> Compiler<CM, SSB> {
         ast: &astCircuit<F, TraceArgs>,
         unit: &mut CompilationUnit<F>,
     ) {
-        for forward_signal in ast.exposed.clone() {
-            let forward_placement = unit.placement.get_forward_placement(&forward_signal);
-            let exposed = (forward_placement.column, forward_placement.rotation);
+        for (queriable, offset) in &ast.exposed {
+            let exposed = match queriable {
+                Queriable::Forward(forward_signal, _) => {
+                    let placement = unit.placement.get_forward_placement(forward_signal);
+                    match offset {
+                        ExposeOffset::First => (placement.column, placement.rotation),
+                        ExposeOffset::Last => {
+                            let rot = placement.rotation + ((unit.num_steps - 1) as i32) * (unit.placement.first_step_height() as i32);
+                            (placement.column, rot)
+                        },
+                        ExposeOffset::Step(step) => {
+                            let rot = placement.rotation + (*step as i32) * (unit.placement.first_step_height() as i32);
+                            (placement.column, rot)
+                        }
+                    }
+                }
+                Queriable::Shared(shared_signal, _) => {
```

Comment:
@lancenonce the shared signal is still with the wrong values for rotation

Reply:
Fixed, thanks @leolara !